File: scripts/hygiene/workflow-shape.mjs
#!/usr/bin/env bun
import { readFileSync, existsSync, readdirSync } from "node:fs";
import { join } from "node:path";
import { parse } from "yaml";

const args = process.argv.slice(2);
const rootIdx = args.indexOf("--root");
const root = rootIdx === -1 ? process.cwd() : args[rootIdx + 1];

const wfDir = join(root, ".github", "workflows");
const violations = [];

if (existsSync(wfDir)) {
  for (const entry of readdirSync(wfDir)) {
    if (!/\.ya?ml$/.test(entry)) continue;
    const path = join(wfDir, entry);
    let doc;
    try {
      doc = parse(readFileSync(path, "utf8"));
    } catch (e) {
      violations.push(`${entry}: could not parse YAML (${e.message})`);
      continue;
    }
    // Article VII.4: workflows declare least-privilege permissions. Either a
    // top-level block, or every job declares its own — either is sufficient,
    // but declaring none at all (inheriting the default token's broad scope)
    // is the violation.
    const jobs = doc?.jobs ?? {};
    if (!("permissions" in (doc ?? {}))) {
      const jobsMissingPermissions = Object.entries(jobs)
        .filter(([, job]) => !("permissions" in (job ?? {})))
        .map(([name]) => name);
      if (jobsMissingPermissions.length > 0) {
        violations.push(`${entry}: no top-level \`permissions:\`, and job(s) ${jobsMissingPermissions.join(", ")} declare none either`);
      }
    }
    for (const [jobName, job] of Object.entries(jobs)) {
      // A job whose entire content is a reusable `uses:` is exempt (its ref is
      // verified separately by `just check-pins`).
      const jobKeys = Object.keys(job ?? {});
      if (jobKeys.length === 1 && jobKeys[0] === "uses") continue;

      const steps = job?.steps ?? [];
      for (const step of steps) {
        if (!("run" in step)) continue; // a `uses:` step (checkout, setup, etc.) is fine
        const stepKeys = Object.keys(step).filter((k) => k !== "env" && k !== "name" && k !== "if" && k !== "run" && k !== "shell" && k !== "working-directory");
        if (stepKeys.length > 0) {
          violations.push(`${entry}: job "${jobName}" step "${step.name ?? "(unnamed)"}" has unexpected keys: ${stepKeys.join(", ")}`);
        }
        const run = String(step.run ?? "").trim();
        if (!/^just\s+\S/.test(run)) {
          violations.push(`${entry}: job "${jobName}" step "${step.name ?? "(unnamed)"}" runs "${run}", which is not \`just <recipe>\``);
        }
      }
    }
  }
}

if (violations.length > 0) {
  console.error("workflow-shape: FAIL");
  for (const v of violations) console.error(`  ${v}`);
  process.exit(1);
}
console.log("workflow-shape: PASS");

File: scripts/hygiene/workflow-shape.py
#!/usr/bin/env python3
import os
import re
import sys

import yaml

args = sys.argv[1:]
root = args[args.index("--root") + 1] if "--root" in args else os.getcwd()

WF_DIR = os.path.join(root, ".github", "workflows")
violations = []

STEP_ALLOWED_KEYS = {"env", "name", "if", "run", "shell", "working-directory"}

if os.path.isdir(WF_DIR):
    for entry in os.listdir(WF_DIR):
        if not re.search(r"\.ya?ml$", entry):
            continue
        path = os.path.join(WF_DIR, entry)
        try:
            with open(path, "r", encoding="utf-8") as f:
                doc = yaml.safe_load(f)
        except Exception as e:
            violations.append(f"{entry}: could not parse YAML ({e})")
            continue

        jobs = (doc or {}).get("jobs") or {}
        if "permissions" not in (doc or {}):
            jobs_missing_permissions = [name for name, job in jobs.items() if "permissions" not in (job or {})]
            if jobs_missing_permissions:
                violations.append(
                    f"{entry}: no top-level `permissions:`, and job(s) {', '.join(jobs_missing_permissions)} declare none either"
                )

        for job_name, job in jobs.items():
            job_keys = list((job or {}).keys())
            if len(job_keys) == 1 and job_keys[0] == "uses":
                continue

            steps = (job or {}).get("steps") or []
            for step in steps:
                if "run" not in step:
                    continue
                step_keys = [k for k in step.keys() if k not in STEP_ALLOWED_KEYS]
                if step_keys:
                    violations.append(
                        f'{entry}: job "{job_name}" step "{step.get("name", "(unnamed)")}" has unexpected keys: {", ".join(step_keys)}'
                    )
                run = str(step.get("run") or "").strip()
                if not re.match(r"^make\s+\S", run):
                    violations.append(
                        f'{entry}: job "{job_name}" step "{step.get("name", "(unnamed)")}" runs "{run}", which is not `make <target>`'
                    )

if violations:
    print("workflow-shape: FAIL", file=sys.stderr)
    for v in violations:
        print(f"  {v}", file=sys.stderr)
    sys.exit(1)
print("workflow-shape: PASS")
